src/microbe_stage/Microbe.Contact.cs

```diff
@@ -1308,51 +1654,229 @@ private void BeginBind()
+        var engulfable = engulfed.Object.Value;
+        if (engulfable == null)
+            return;
 
-        microbe.hostileEngulfer.Value = null;
+        engulfable.PhagocytosisStep = PhagocytosisPhase.Ingested;
+
+        attemptingToEngulf.Remove(engulfable);
+        touchedEntities.Remove(engulfable);
+
+        UsedIngestionCapacity += engulfable.EngulfSize;
+        engulfed.HasBeenIngested = true;
+
+        OnSuccessfulEngulfment?.Invoke(this, engulfable);
+        engulfable.OnIngestedFromEngulfment();
+    }
+
+    private void CompleteEjection(EngulfedObject engulfed)
+    {
+        var engulfable = engulfed.Object.Value;
+        if (engulfable == null)
+            return;
+
+        attemptingToEngulf.Remove(engulfable);
+        engulfedObjects.Remove(engulfed);
+        expelledObjects.Add(engulfed);
+
+        engulfed.ReclaimedByAnotherHost = false;
+        engulfable.PhagocytosisStep = PhagocytosisPhase.None;
+
+        foreach (string group in engulfed.OriginalGroups)
+        {
+            if (group != Constants.RUNNABLE_MICROBE_GROUP)
+                engulfable.EntityNode.AddToGroup(group);
+        }
+
+        // Reset render priority
+        engulfable.RenderPriority = engulfed.OriginalRenderPriority;
+
+        engulfed.Phagosome.Value?.DestroyDetachAndQueueFree();
+
+        // Ignore possible invalid cast as the engulfed node should be a rigidbody either way
+        var body = (RigidBody)engulfable;
+
+        body.Mode = ModeEnum.Rigid;
+
+        // Reparent to world node
+        body.ReParentWithTransform(GetStageAsParent());
+
+        // Set to default microbe collision layer and mask values
+        body.CollisionLayer = 3;
+        body.CollisionMask = 3;
+
+        var impulse = Transform.origin.DirectionTo(body.Transform.origin) * body.Mass *
+            Constants.ENGULF_EJECTION_FORCE;
+
+        // Apply outwards ejection force
+        body.ApplyCentralImpulse(impulse + LinearVelocity);
+
+        // We have our own engulfer and it wants to claim this object we've just expelled
+        HostileEngulfer.Value?.IngestEngulfable(engulfable);
+
+        engulfable.OnExpelledFromEngulfment();
+        engulfable.HostileEngulfer.Value = null;
+    }
+
+    /// <summary>
+    ///   Stores extra information to the objects that have been engulfed.
+    /// </summary>
+    private class EngulfedObject
+    {
+        [JsonConstructor]
+        public EngulfedObject(IEngulfable @object, Endosome endosome)
+        {
+            Object = new EntityReference<IEngulfable>(@object);
+            Phagosome = new EntityReference<Endosome>(endosome);
+
+            AdditionalEngulfableCompounds = @object.CalculateAdditionalDigestibleCompounds()?
+                .Where(c => c.Key.Digestible)
+                .ToDictionary(c => c.Key, c => c.Value);
+
+            InitialTotalEngulfableCompounds = @object.Compounds.Compounds
+                .Where(c => c.Key.Digestible)
+                .Sum(c => c.Value);
```

Comment:
Here's my final review comment: shouldn't these be JSON saved, because otherwise the initial things to engulf will be incorrect when loading a save? So I'd actually create a different constructor that has these as parameters, and mark that as the json constructor.

Reply:
Not sure if I get what you mean, shouldn't these properties already be saved even without `[JsonProperty]`? Also by another constructor I assume it would be like this?
```csharp
        [JsonConstructor]
        public EngulfedObject(Dictionary<Compound, float> additionalEngulfableCompounds,
            float initialTotalEngulfableCompounds)
        {
            AdditionalEngulfableCompounds = additionalEngulfableCompounds;
            InitialTotalEngulfableCompounds = initialTotalEngulfableCompounds;
        }
```

Reply:
Huh, surprising it seems to work fine all this time.

Anyways, I'm thinking of ditching constructors and just  use object initializer all the way.

Reply:
Well the constructor would recreate the data. So the end result is that the original compounds left gets reinitialized to the current compounds left, meaning that the progress from 0 to fully digested, would get reset when loading a save. That'd probably only be visible when the player is engulfed, and otherwise completely unnoticeable.
In this case as the created data is dependent on the constructor parameter, I think it's a very good case for having a constructor.